com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java

```diff
@@ -23,29 +32,119 @@
  * 
  */
 public class ViatraQueryAdapter extends AdapterImpl {
+	public static final String MESSAGE_ENGINE_NOT_READY = 
+			"Cannot initialize VIATRA Query Engine until the project is loaded.";
+	private static final Logger LOGGER = Logger.getLogger(ViatraQueryAdapter.class);
 	
-	private AdvancedViatraQueryEngine engine;
+	private Optional<AdvancedViatraQueryEngine> engine = Optional.empty();
 	private Project project;
 	private boolean engineDisposable = true;
 	private Set<String> identifierMap = new HashSet<>();
+	private boolean isInitialized = false;
+	private Notifier[] notifiers;
+	private List<Consumer<AdvancedViatraQueryEngine>> initializationActions = new ArrayList<>();
 
-	private ViatraQueryAdapter(AdvancedViatraQueryEngine engine, Project project) {
-		this.engine = engine;
+	private ViatraQueryAdapter(Project project, Notifier[] notifiers) {
 		this.project = project;
+		this.notifiers = notifiers;
 	}
-	
+
+	/**
+	 * @return initialized engine
+	 * @throws ViatraQueryException if engine cannot be initialized yet
+	 * @deprecated Use {@link #getInitializedEngineChecked()} or {@link #getInitializedEngine()} instead
+	 */
+	@Deprecated
 	public AdvancedViatraQueryEngine getEngine() {
+		return getInitializedEngineChecked();
+	}
+
+	/**
+	 * Try to initialize the VIATRA engine if it hasn!t been before and if the 
+	 * initialization is not possible, it throws a ViatraQueryException.
+	 * 
+	 * @return the initialized engine
+	 * @throws ViatraQueryException if initialized engine is not available yet
+	 */
+	public AdvancedViatraQueryEngine getInitializedEngineChecked() {
```

Comment:
Why is this a separate public method? 
 - Does exactly the same as getEngine
 - Newly introduced so serves no purpose in backwards compatibility

What I'm guessing is to have an equivalent of the getEngine method following the new naming conventions, however imho it just makes the API confusing

If the purpose is to replace the current getEngine, we should specify ViatraQueryException in a proper throws block --> I know it is a runtime exception but the user needs to handle it from the calling code.

Reply:
It was me who suggested this to Tamás, and my reasoning was to have a consistent terminology of accessing the engines, and the old getEngine really did not fit. As you have noticed, the new method is aimed to be a clear replacement for the old one, but we did not want to break all existing code (hence we only deprecated the getEngine method).

About adding the throws declaration, I am not against it, but it really does not add a lot, and RuntimeExceptions are not expected to be added as a throws declaration.

Reply:
Would it be better if we handle the exception in the original `getEngine` and return `null` instead of throwing an exception?

Reply:
I don't think so. Returning null is a pattern we very much try to avoid here, especially in case of errors. Here the exception has a lot of added value: we explicitly say that something went wrong instead of the user figuring it out later when the engine got corrupted at a random model update.

Reply:
I think throwing a non-runtime exception would solve all of my concerns. I agree though that the circumstances are pretty much the definition for runtime exceptions, but I think the specific nature of the platform necessitates that the user of this API always handles the cases when the engine is not ready

Right now when migrating or using the non-optional method (which the user will do cause its simpler) the user is not enforced to handle situations where the engine is not ready. And in most cases in MD uncaught exceptions coming from certain integrated functionality can break the entire client. I'd rather have the users add some unnecesary try catch blocks.

Reply:
Also it is important to know that the exception is not thrown when the engine is corrupted due to irregularities, but rather when the user tries to use an engine before it is ready.

Reply:
The problem is that adding a non-runtime exception only to new methods would ensure users have to check for them. This would be an outlier for all existing APIs where it was a design decision to follow VIATRA's approach of always throwing runtime exceptions (and changing all the APIs would break compatibility too much).

On the other hand, I have no problem adding documentation to the Javadoc, stating that this method should only be used by users who can be sure that the project loading is finished. Otherwise, the user should rely on the method returning an optional. If you have a name idea that better suggests this constraint, I am OK with renaming it as well. Would such a compromise solve your concerns?

Reply:
What do you think about throwing a non-runtime exception? In our case, we know the problem, it can be handled by time and there is nothing fatal from which we cannot stand up. Maybe we can have a custom `UnsuccessfulViatraEngineInitializationException` or something like this.

Reply:
This is only a minor concern though I'm willing to let it go if it's documented properly. (which seems to be the case)

Reply:
Based on the discussion above (and because I have found a typo) I suggest rewriting the Javadoc as follows (and I cannot really add it as an inline recommendation because it is split into multiple lines and I would like to add it in a reviewable text here (please check the Javadoc syntax, I was writing it as a text):

```
Returns an initialized VIATRA query engine stored in the adapter.
If a query engine was initialized earlier, that engine will be returned;
if no query engine was initialized earlier, a new one is created. If
the initialization fails, e.g. the engine was requested before the
model was loaded entirely, a {@link ViatraQueryException} is thrown.
</p>
This method should only be called if the engine initialization is
expected to be completed successfully, otherwise it is recommended
to use {@link #getInitializedEngine()},
{@link #getInitializedEngine(Consumer<AdvancedViatraQueryEngine>)}
or {@link #requireMatcher(IQuerySpecification<M>)} instead
as they are aimed to handle the initialization issues better.

@return the initialized engine
@throws ViatraQueryException if no query engine was initialized before and initialization failed
```